lib/elixir/src/elixir_expand.erl

```diff
@@ -733,6 +734,16 @@ allowed_in_context(_, _Arity, guard) ->
 allowed_in_context(_, _, _) ->
   true.
 
+assert_valid_args_for_remote({{'.', _, [erlang, '++']}, Meta, [{_, _} = Left, _Right]}, E) ->
```

Comment:
Hrm, I think we are being a bit inconsistent here. For example, this code warns at best:

```elixir
1 + :foo
```

While being obviously invalid. I wouldn't want to add all of those checks to the compiler, since they are already performed by a later Erlang pass anyway.

There are other places, such as guards, where doing so is weird, although semantically valid:

```elixir
def foo when 1 ++ 2 do
  ...
end
```

So my suggestion is for us to raise when the context is a match and whenever the left side is not a list. This will simplify the check, make sure we don't raise an internal erlang error, and make `++` behave like the other operators.

Reply:
I did some tests and I believe is not valid semantically inside guards:
```
iex(1)> defmodule A do
...(1)> def foo when 1 ++ 2 do
...(1)> 1
...(1)> end
...(1)> end
** (CompileError) iex:2: illegal guard expression
    (stdlib) lists.erl:1338: :lists.foreach/2
    (stdlib) erl_eval.erl:670: :erl_eval.do_apply/6
    (iex) lib/iex/evaluator.ex:245: IEx.Evaluator.handle_eval/5
```

But it's also invalid in Erlang and doesn't impact the solution. I agree that is simpler in the matcher only, and let Erlang handle the other errors, I just don't know if I understood what you mean with changes for `<<>>`, but I think you removed.